### PR TITLE
Replicating APIStoreHostObject and APIPublisherHostObject methods in APIProviderImpl and APIUtil.java

### PR DESCRIPTION
**Pull request by thushara35**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/CommonUtil.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        }
+    }
+
+    /**
+     * Validate the backend by sending HTTP HEAD
+     *
+     * @param urlVal - backend URL
+     * @return - status of HTTP HEAD Request to backend
+     */
+    public static String sendHttpHEADRequest(String urlVal) {
```
Comment (shazni) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/CommonUtil.java]: Not sure why returning String to check for connectivity. Any specific reason? Isn't returning a boolean work here?

Comment (thushara35) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/CommonUtil.java]: we returning a string value for use in the front-end, actual use-case used in a worst case scenario to know what went wrong.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 2]
```diff
@@ -3104,7 +3113,90 @@ private String saveAPI(API api,
...
+                PrivilegedCarbonContext.endTenantFlow();
+            }
+        }
+        return apiOlderVersionExist;
+    }
+
+    /**
+     * This method check whether URL valid or not
+     *
+     * @param type
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Add descriptions for the parameters. Mention that when could the exception is thrown.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 3]
```diff
@@ -136,6 +144,24 @@
     private static Set<Integer> registryInitializedTenants = new HashSet<Integer>();
     private static GenericArtifactManager genericArtifactManager;
 
+    private static ConfigurationContextService configContextService = null;
+
+    public static void setConfigContextService(ConfigurationContextService configContext) {
```
Comment (prasa7) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: Please add method comments. Check the other places as well.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 4]
```diff
@@ -3962,4 +3988,677 @@ public static void updatePermissionCache(String username) throws UserStoreExcept
...
+        String backendPort;
+        try {
+            port = CarbonUtils.getTransportProxyPort(getConfigContext(), transport);
+            if (port == -1) {
+                port = CarbonUtils.getTransportPort(getConfigContext(), transport);
+            }
+            backendPort = Integer.toString(port);
+            return backendPort;
+        } catch (APIManagementException e) {
+            log.error(e.getMessage());
```
Comment (pulasthi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: you also need to send "e"  -- log.error(e.getMessage(),e);

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/CommonUtil.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            //If the endpoint doesn't support HTTP HEAD or if status code is < 400
+            if (statusCode == 405 || statusCode % 100 < 4) {
+                if (log.isDebugEnabled() && statusCode == 405) {
+                    log.debug("Endpoint doesn't support HTTP HEAD");
+                }
+                response = "success";
+            }
+        } catch (IOException e) {
+            // sending a default error message.
+            log.error("Error occurred while connecting backend : " + urlVal + ", reason : " + e.getMessage());
```
Comment (pulasthi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/CommonUtil.java]: send "e" with the log.error

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 2 above)
Comment (thushara35) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: This method is being removed
Thanks,

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java (hunk 3 above)
Comment (thushara35) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: PR sent : https://github.com/wso2/carbon-apimgt/pull/577

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java (hunk 4 above)
Comment (thushara35) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: PR sent : https://github.com/wso2/carbon-apimgt/pull/577

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/CommonUtil.java (hunk 5 above)
Comment (thushara35) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/CommonUtil.java]: PR sent : #577

**Pull request merged**